src/dune_engine/scheduler.ml

```diff
@@ -759,45 +782,62 @@ let maybe_clear_screen ~config =
          ])
 
 let poll ?config ~once ~finally () =
-  let t = prepare ?config () in
-  let watcher = File_watcher.create () in
-  let block_waiting_for_changes () =
-    match Event.next () with
-    | Job_completed _ -> assert false
-    | Files_changed -> Continue
-    | Signal signal ->
-      got_signal signal;
-      Exit
-  in
-  let wait msg =
-    Console.Status_line.set_temporarily
-      (fun () ->
-        Some (Pp.seq msg (Pp.verbatim ", waiting for filesystem changes...")))
-      (fun () ->
-        let res = block_waiting_for_changes () in
-        maybe_clear_screen ~config;
-        res)
+  let t = prepare ?config ~polling:true () in
+  let finally () =
+    (* Re-create the throttle so that we don't keep old ivars from one run to
+       the next *)
+    t.job_throttle <- Fiber.Throttle.create (Fiber.Throttle.size t.job_throttle);
+    Fiber.return (finally ())
   in
+  let watcher = File_watcher.create () in
   let rec loop () =
-    let res = Run_once.run_and_cleanup t once in
-    finally ();
-    match res with
+    t.status <- Building;
+    let open Fiber.O in
+    let* res =
+      let on_error exn () =
+        match t.status with
+        | Building -> Report_error.report exn
+        | Cancelling_build -> ()
+        | Waiting_for_file_changes _ ->
+          (* We are inside a build, so we aren't waiting for a file change event *)
+          assert false
+      in
+      Fiber.fold_errors ~init:() ~on_error (fun () ->
+          Fiber.finalize ~finally once)
+    in
+    match t.status with
+    | Waiting_for_file_changes _ ->
+      (* We just finished a build, so there's no way this was set *)
+      assert false
+    | Cancelling_build -> loop ()
```

Comment:
My first thought here: "This is weird. Why are we building if we're told to cancel?".
But of course for us "cancelling build" means "restarting build", which means the code is correct. Do you think Restarting_build would be a better name? (or is there an even better one)

Reply:
Good call. I'll rename.